File: spring-webflux-java/src/main/java/com/example/springwebfluxjava/handler/ApiHandler.java
package com.example.springwebfluxjava.handler;

import com.example.springwebfluxjava.document.Movie;
import com.example.springwebfluxjava.dto.JikanResponse;
import com.example.springwebfluxjava.repository.MovieRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
@AllArgsConstructor
public class ApiHandler {
    private final MovieRepository movieRepository;
    private WebClient client;

    public Mono<ServerResponse> getAnimeAndMovie(ServerRequest req) {
        var id = req.pathVariable("id");
        var anime = client.get()
                .uri("https://api.jikan.moe/v4/anime/%s".formatted(id))
                .retrieve()
                .bodyToMono(JikanResponse.class)
                .flatMap(it -> Mono.just(it.getAnimeDto()));

        var movie = movieRepository.findOneRandomMovie();

        return ServerResponse.ok().body(Flux.merge(anime, movie), Flux.class);
    }

    public Mono<ServerResponse> getMovies(ServerRequest req) {
        var count = req.queryParam("count");

        int countVal;
        try {
            countVal = count.map(Integer::parseInt).orElse(10);
        } catch (NumberFormatException e) {
            return ServerResponse.badRequest().build();
        }

        if (countVal > 0 && countVal <= 25) {
            var movies = movieRepository.findRandomMovies(countVal);
            return ServerResponse.ok().body(movies, Movie.class);
        }

        return ServerResponse.badRequest().build();
    }
}

File: spring-webflux-kotlin/src/main/kotlin/com/example/springwebfluxkotlin/handler/ApiHandler.kt
package com.example.springwebfluxkotlin.handler

import com.example.springwebfluxkotlin.dto.ResponseData
import com.example.springwebfluxkotlin.repository.MovieRepository
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.merge
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.server.*

@Component
class ApiHandler(
    private val movieRepository: MovieRepository,
    private val client: WebClient
) {

    suspend fun getAnimeAndMovie(req: ServerRequest) = coroutineScope {
        val id = req.pathVariable("id")
        val anime = async {
            client.get()
                .uri("https://api.jikan.moe/v4/anime/$id")
                .accept(MediaType.APPLICATION_JSON)
                .retrieve()
                .awaitBody<ResponseData>()
                .data
        }

        val movie = async {
            movieRepository.findOneRandomMovie()
        }

        ServerResponse.ok()
            .json()
            .bodyAndAwait(merge(flowOf(anime.await()), movie.await()))
    }

    suspend fun getMovies(req: ServerRequest): ServerResponse {
        val count = try {
            req.queryParamOrNull("count")?.toInt() ?: 10
        } catch (e: NumberFormatException) {
            return ServerResponse.badRequest().buildAndAwait()
        }

        return if (count in 1..25) {
            ServerResponse.ok()
                .json()
                .bodyAndAwait(movieRepository.findRandomMovies(count))
        } else {
            ServerResponse.badRequest().buildAndAwait()
        }
    }

}